### DISCUSSION TITLE
Support `AsEnumCollection` cast

### DISCUSSION THREAD

**Discussion by susanu** [Laravel-Backpack/community-forum#1258] [Ideas]
I have one attribute that stores multiple enum values.
Currently there is no field/column that support this.
Do you think is good addition to the current list of fields & columns?

[a comment by an automated account is not shown]

**Comment by pxpm** [comment]
@susanu thanks. 🙏 

What do you mean multiple enum values? A JSON in the database? `{type: value1, brand: value2, color: value3 }` ? It looks like the repeatable could handle that with `max_rows => 1` and `min_rows => 1` ? 

If not can you show me a bit more of what you mean ? 

Cheers

**Comment by susanu** [comment]
The data is saved into the database as a json array, when the model is loaded, is transformed into a Collection with enums (the class)

SmsScope.php
```
<?php

enum SmsScope: string
{
    case AbandonedCart = 'abandoned_cart';
    case Campaign = 'campaign';
}
```

Model (casts function)
```
['scopes' => Illuminate\Database\Eloquent\Casts\AsEnumCollection::of(SmsScope::class)];
```

Database
```
["campaign","abandoned_cart"]
```
OR
```
["campaign"]
```

Column
```
CRUD::column('scopes')
    ->type('text')
    ->value(fn($entry) => $entry->scopes->implode(fn(SmsScope $s) => $s->name, ', '));
```

Currently i use the field type ```select2_from_array``` but i transform the attribute from Collection to normal array when the model is retrieved.

**Reply by pxpm** [reply]
Thanks for clarifying @susanu 🙏 

Ok, I think it's a reasonable request to implement since we already support single enum values. Any chance this could be dealt with `enum` column/field we currently have in a conditional? 

My idea is that `enum` could support both the `Enum` cast and the `AsEnumCollection` so you don't need to remember and/or define different column types/field types for each of the enums. It's an enum? Use the `enum` column/field. 

What do you think ? 

Cheers

**Reply by susanu** [reply]
Yes, you are totally right.
You can target collection of enums by checking the cast if is AsEnumCollection.

Thank you!

**Reply by pxpm** [reply]
Hey @susanu 

We've added support for `AsEnumCollection` cast. 

Let me know if that works for you 👍

**Comment by mattsches** [comment]
I've just upgraded to 6.7.48 (from 6.7.43), and I run in to an error with enums. I haven't looked at it any closer now, but it happens when calling the `columns/enum.blade.php` view, and the error message is `Call to a member function transform() on string`;

I'll let you know once I find out more.

**Reply by pxpm** [reply]
Will investigate too, thanks 👍

**Reply by pxpm** [reply]
@mattsches released a new crud version with the fix. Let me know if you are still experiencing problems. 👍

**Reply by mattsches** [reply]
@pxpm Thanks for the quick reply and fix.
I actually fixed my issue by defining the `enum_class` in the column/field configs - which I hadn't previously (but it still had worked). I also found that my actual Enum class lacked `case`s and only had `constant`s, so I changed that, too. These two changes seem to have fixed the issue for me with `6.7.48` - but I'll update to `6.7.49` and check if it works there, too.

**Reply by mattsches** [reply]
Ok, still looking good after update to `6.7..49`.

I also tried and reverted my changes and tested the old version of my code against `6.7.49` - works, too! So, good job :+1:
